MorphicServer/UnregisterEndpoint.cs

```diff
@@ -0,0 +1,105 @@
+// compliance with this License.
+//
+// You may obtain a copy of the License at
+// https://github.com/GPII/universal/blob/master/LICENSE.txt
+//
+// The R&D leading to these results received funding from the:
+// * Rehabilitation Services Administration, US Dept. of Education under 
+//   grant H421A150006 (APCP)
+// * National Institute on Disability, Independent Living, and 
+//   Rehabilitation Research (NIDILRR)
+// * Administration for Independent Living & Dept. of Education under grants 
+//   H133E080022 (RERC-IT) and H133E130028/90RE5003-01-00 (UIITA-RERC)
+// * European Union's Seventh Framework Programme (FP7/2007-2013) grant 
+//   agreement nos. 289016 (Cloud4all) and 610510 (Prosperity4All)
+// * William and Flora Hewlett Foundation
+// * Ontario Ministry of Research and Innovation
+// * Canadian Foundation for Innovation
+// * Adobe Foundation
+// * Consumer Electronics Association Foundation
+
+using System.Threading.Tasks;
+using MorphicServer.Attributes;
+using System.Text.Json.Serialization;
+using Serilog;
+
+namespace MorphicServer
+{
+    public abstract class UnregisterEndpoint : Endpoint
+    {
+        protected async Task Unregister<TCredential>(User user, TCredential cred) where TCredential : Record
+        {
+            var db = Context.GetDatabase();
+            var deleted = await db.DeleteMany<Preferences>(r => r.UserId == user.Id);
+            Log.Logger.Debug($"Deleted {deleted} Preferences");
+            deleted = await db.DeleteMany<AuthToken>(r => r.UserId == user.Id);
+            Log.Logger.Debug($"Deleted {deleted} AuthTokens");
+            deleted = await db.DeleteMany<BadPasswordLockout>(r => r.Id == user.Id);
+            Log.Logger.Debug($"Deleted {deleted} AuthTokens");
+            await Delete(cred);
+            await Delete(user);
+        }
+    }
+    
+    public class UnregisterUsernameRequest
+    {
+        [JsonPropertyName("username")]
+        public string Username { get; set; } = "";
+
+        [JsonPropertyName("password")]
+        public string Password { get; set; } = "";
+    }
+
+    public class UnregisterKeyRequest
+    {
+        [JsonPropertyName("key")]
+        public string Key { get; set; } = "";
+    }
+
+
+    /// <summary>
+    /// Endpoint Controller for username unregister requests. Security relies on the username-password
+    /// being passed in the request. The reason is two-fold: 1) the user may decide to unregister separately
+    /// from using the product (and perhaps not having logge in in a while), so it seems weird to ask them
+    /// to log in first, and then unregister (though perhaps that's not so unreasonable), and 2) since this
+    /// is a destructive action we don't want someone who hijacks the auth-token to be able to do this.
+    ///
+    /// TODO We could make this require the auth token, but I would still recommend we make the user enter their password for extra security anyway
+    /// </summary>
+    [Path("/v1/unregister/username")]
```

Comment:
Shouldn't it be something like `/v1/users/{userId}/unregister`?  If they're deleting their account, it doesn't really matter what credentials they use.  I probably disagree that it seems weird to ask them to log in first.  For point #2, is there a way to require a new auth token?  Maybe we could add some kind of `Use` to `AuthToken`, and the client can take care of re-authenticating however it makes sense for the particular user, requesting in the auth POST that the new `AuthToken` have a `Use` of `"Unregister"`.

Reply:
Isn't that really a one-time-auth-token, then? I can make this another email-thing, where we mail an unregister-one-time-link to the user that they click on. One could argue that makes is harder to unregister, but perhaps want to make it a little hard, so we keep our users (though I'm not sure ethically where that falls).